netty/src/main/java/io/grpc/netty/NettyServer.java

```diff
@@ -258,51 +297,101 @@ public void operationComplete(ChannelFuture future) throws Exception {
+                ChannelFuture future = b.bind(address);
+                channelGroup.add(future.channel());
+                final InternalInstrumented<SocketStats> listenSocketStats =
+                    new ListenSocket(future.channel());
+                future.addListener(new ChannelFutureListener() {
+                  @Override
+                  public void operationComplete(ChannelFuture future) throws Exception {
+                    channelz.addListenSocket(listenSocketStats);
+                    listenSocketStatsList.add(listenSocketStats);
+                  }
+                });
+                ChannelFutureListener closeListener = new ChannelFutureListener() {
+                  @Override
+                  public void operationComplete(ChannelFuture future) throws Exception {
+                    if (!future.isSuccess()) {
+                      log.log(Level.WARNING, "Error closing server channel", future.cause());
+                    }
+                    channelz.removeListenSocket(listenSocketStats);
+                    listenSocketStatsList.remove(listenSocketStats);
+                  }
+                };
+                future.channel().closeFuture().addListener(closeListener);
+                bindFutures.put(address, ImmutableMap.of(future, closeListener));
+            }
+            return bindFutures;
+          }
+        }
+    );
+    Map<SocketAddress, Map<ChannelFuture, ChannelFutureListener>> channelFutures =
+        bindCallFuture.awaitUninterruptibly().getNow();
+
+    if (!bindCallFuture.isSuccess()) {
+      channelGroup.close().awaitUninterruptibly();
+      throw new IOException(String.format("Failed to bind to addresses %s",
+          addresses), bindCallFuture.cause());
     }
-    channel = future.channel();
-    channel.eventLoop().execute(new Runnable() {
-      @Override
-      public void run() {
-        listenSocketStats = new ListenSocket(channel);
-        channelz.addListenSocket(listenSocketStats);
+    for (Map.Entry<SocketAddress, Map<ChannelFuture, ChannelFutureListener>> entry:
+        channelFutures.entrySet()) {
+      // We'd love to observe interruption, but if interrupted we will need to close the channel,
+      // which itself would need an await() to guarantee the port is not used when the method
+      // returns. See #6850
+      for (ChannelFuture future: entry.getValue().keySet()) {
+        if (!future.awaitUninterruptibly().isSuccess()) {
+          channelGroup.close().awaitUninterruptibly();
+          throw new IOException(String.format("Failed to bind to address %s",
+              entry.getKey()), future.cause());
+        }
       }
-    });
+    }
+    for (Map<ChannelFuture, ChannelFutureListener> channelListener: channelFutures.values()) {
+      for (Map.Entry<ChannelFuture, ChannelFutureListener> entry: channelListener.entrySet()) {
+        entry.getKey().channel().closeFuture().removeListener(entry.getValue());
```

Comment:
Why do this? If you just leave the listener in-place, then you don't have to communicate it back to this thread (in the awkward map) and you don't need to call removeListenSocket within shutdown().

Reply:
The close listener installed on the channel will stay with it in the entire lifetime, so if we do not remove it, then it conflicts with removeListenSocket at shutdown(), isn't it?

Reply:
I agree, but it seems simpler to remove the removal in shutdown().

Reply:
I know where the conflict was from. Left the individual channel close listener removal there and the group close listener removal is ok to run again because of the emptiness check.

Reply:
But you _don't need_ the group close listener removal! And setting `listenSocketStatsList = null` doesn't add much value and makes the code more complex. Why should this code even exist if the other code will handle the cleanup?